cpp/src/decisiontree/levelalgo/levelkernel_classifier.cuh

```diff
@@ -0,0 +1,274 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#include "common_kernel.cuh"
+#include "cub/cub.cuh"
+
+__global__ void gini_kernel_level(const int* __restrict__ labels,
+                                  const unsigned int* __restrict__ sample_cnt,
+                                  const int nrows, const int nmax,
+                                  int* histout) {
+  int threadid = threadIdx.x + blockIdx.x * blockDim.x;
+  extern __shared__ unsigned int shmemhist[];
+  if (threadIdx.x < nmax) shmemhist[threadIdx.x] = 0;
```

Comment:
This assumes that the number of unique labels will always be less than or equal to `blockDim.x`. Is this a valid assumption to make?

Reply:
That would be true in most cases. But i will put a loop around. Fixed

Reply:
yes, please. Loop would make this more generic.